sw/device/tests/otbn/otbn_rsa_test.c

```diff
@@ -0,0 +1,903 @@
+                          const uint8_t *private_exponent, const uint8_t *in,
+                          const uint8_t *encrypted_expected,
+                          uint8_t *out_encrypted, uint8_t *out_decrypted) {
+  dif_otbn_config_t otbn_config = {
+      .base_addr = mmio_region_from_addr(TOP_EARLGREY_OTBN_BASE_ADDR),
+  };
+
+  otbn_rsa_t otbn_rsa_ctx;
+
+  // Initialize
+  profile_start();
+  CHECK(otbn_rsa_init(&otbn_rsa_ctx, &otbn_config) == kOtbnRsaOk);
+  CHECK(otbn_rsa_load_app(&otbn_rsa_ctx) == kOtbnRsaOk);
+  profile_end("Initialization");
+
+  // Encrypt
+  LOG_INFO("Encrypting");
+  profile_start();
+  CHECK(otbn_rsa_encrypt(&otbn_rsa_ctx, modulus, in, out_encrypted,
+                         size_bytes) == kOtbnRsaOk);
+  check_data(out_encrypted, encrypted_expected, size_bytes);
+  profile_end("Encryption");
+
+  if (kTestDecrypt) {
+    // Decrypt
+    LOG_INFO("Decrypting");
+    profile_start();
+    CHECK(otbn_rsa_decrypt(&otbn_rsa_ctx, modulus, private_exponent,
+                           encrypted_expected, out_decrypted,
+                           size_bytes) == kOtbnRsaOk);
+    check_data(out_decrypted, in, size_bytes);
+    profile_end("Decryption");
+  }
+}
+
+static void test_rsa512_roundtrip() {
+  static const uint8_t modulus[kRsa512SizeBytes] = {
+      0xf3, 0xb7, 0x91, 0xce, 0x6e, 0xc0, 0x57, 0xcd, 0x19, 0x63, 0xb9,
+      0x6b, 0x81, 0x97, 0x96, 0x40, 0x81, 0xd8, 0x89, 0x27, 0xec, 0x0a,
+      0xb1, 0xf2, 0x4a, 0xda, 0x2e, 0x68, 0xe1, 0x80, 0xa4, 0x4f, 0xe0,
+      0x82, 0x87, 0x1f, 0x98, 0xdc, 0x42, 0xc7, 0xc2, 0xce, 0xa2, 0xb2,
+      0x1a, 0x3f, 0x77, 0xdc, 0xc6, 0x27, 0x6d, 0x83, 0x5c, 0xcd, 0x1d,
+      0xdf, 0xe5, 0xf3, 0x98, 0xe6, 0x8b, 0xe6, 0x5b, 0xd4};
+
+  static const uint8_t private_exponent[kRsa512SizeBytes] = {
+      0xc1, 0xf3, 0x5d, 0x18, 0x12, 0x7e, 0xe7, 0x0c, 0xbf, 0x33, 0xd0,
+      0x1c, 0xd8, 0x5d, 0x91, 0x26, 0xb6, 0xc5, 0xae, 0x78, 0xda, 0x4c,
+      0xae, 0x43, 0xa1, 0x57, 0xab, 0x32, 0xcf, 0xa4, 0xd4, 0x72, 0x20,
+      0x53, 0x30, 0x55, 0x7a, 0x93, 0xd9, 0xae, 0x75, 0x32, 0x9d, 0x09,
+      0x18, 0x06, 0xc8, 0x26, 0x64, 0x28, 0xcf, 0x2c, 0x3b, 0x6e, 0x6b,
+      0x5c, 0x28, 0xde, 0x76, 0x6c, 0x2f, 0xcc, 0xf3, 0x31};
+
+  static const uint8_t encrypted_expected[kRsa512SizeBytes] = {
+      0xb7, 0x02, 0x28, 0xcb, 0x63, 0x5e, 0xa6, 0xfd, 0x55, 0x4a, 0x85,
+      0x43, 0x1d, 0x26, 0x13, 0xb3, 0x78, 0x66, 0xd9, 0xe2, 0xe1, 0xbf,
+      0x29, 0xc6, 0xc6, 0xdd, 0x90, 0x76, 0x3f, 0x1d, 0x43, 0xc0, 0x76,
+      0x51, 0x75, 0x10, 0x66, 0x61, 0x8c, 0x3c, 0x99, 0xd9, 0x90, 0xd2,
+      0x59, 0x45, 0x0a, 0x7a, 0x6d, 0x58, 0xaa, 0x75, 0xf2, 0x63, 0xb3,
+      0xe1, 0x06, 0x4b, 0x82, 0x0a, 0xdd, 0x07, 0x44, 0x2a};
+
+  static const uint8_t in[kRsa512SizeBytes] = {
+      "Hello OTBN, can you encrypt and decrypt this for me?"};
+  uint8_t out_encrypted[kRsa512SizeBytes] = {0};
+  uint8_t out_decrypted[kRsa512SizeBytes] = {0};
```

Comment:
Presumably the = {0}; is unnecessary if you're just going to write to them?

Reply:
Since this is a test, I find it helpful to see bogus data more easily by initializing with zero.